hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaDeploymentManagement.java

```diff
@@ -289,8 +290,8 @@ private DistributionSetAssignmentResult assignDistributionSetToTargets(final Str
                 .distinct().collect(Collectors.toList());
 
         final List<String> existingTargetIds = Lists.partition(providedTargetIds, Constants.MAX_ENTRIES_IN_STATEMENT)
-                .stream().map(targetRepository::filterNonExistingControllerIds).flatMap(List::stream)
-                .collect(Collectors.toList());
+                .stream().map(ids -> targetRepository.findAll(TargetSpecifications.hasControllerIdIn(ids)))
```

Comment:
As far as I can see we could collect the targets directly instead of collecting the IDs first. The filtering provided by the `assignmentStrategy` could just provide a spec to be included then instead of filtering with additional database calls afterwards.

Reply:
Yes, that is a good idea: combine specifications directly under Online/Offline assignment strategies `findTargetsForAssignment` method.

Reply:
Okay trying to rework it as suggested I think I found the reason for it being separated into two steps. The first result contains the targets the DS should be assigned to, the second filter only the ones that have it not assigned yet. In case it is assigned to all targets already we display the count of the first filter result, when combining the filters this count is lost.

Reply:
Tried to use a `count()` instead which is not sufficient though as the id list is later filtered by another condition so I will keep this split two staged filtering.

Reply:
Please consider custom spring data repository implementation (see https://docs.spring.io/spring-data/jpa/docs/current/reference/html/#repositories.custom-implementations) in order to select only required/specified field.

Reply:
Custom repo is not easily possible as specifications need entity information is only provided during proxy creation (or otherwise would need a bigger effort using reflections)